File: src/main/java/serendustry/machine/MetaTileEntityIndustrialLaboratory.java
package serendustry.machine;

import com.google.common.collect.Table;
import serendustry.machine.LaboratoryProperty.LaboratoryEntry;
import gregtech.api.GTValues;
import gregtech.api.capability.impl.MultiblockRecipeLogic;
import gregtech.api.metatileentity.MetaTileEntity;
import gregtech.api.metatileentity.TieredMetaTileEntity;
import gregtech.api.metatileentity.interfaces.IGregTechTileEntity;
import gregtech.api.metatileentity.multiblock.IMultiblockPart;
import gregtech.api.metatileentity.multiblock.RecipeMapMultiblockController;
import gregtech.api.pattern.BlockPattern;
import gregtech.api.pattern.FactoryBlockPattern;
import gregtech.api.pattern.PatternMatchContext;
import gregtech.api.pattern.TraceabilityPredicate;
import gregtech.api.recipes.Recipe;
import gregtech.api.recipes.RecipeMap;
import gregtech.api.recipes.RecipeMaps;
import gregtech.client.renderer.ICubeRenderer;
import gregtech.client.renderer.texture.Textures;
import gregtech.common.blocks.BlockGlassCasing;
import gregtech.common.blocks.BlockMachineCasing;
import gregtech.common.blocks.MetaBlocks;
import net.minecraft.block.state.IBlockState;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.math.BlockPos;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.util.HashMap;
import java.util.Map;

public class MetaTileEntityIndustrialLaboratory extends RecipeMapMultiblockController {

    private Mode mode = Mode.NORMAL;

    private final LaboratoryResources laboratoryResources = new LaboratoryResources();

    //private final Map<BlockPos, MetaTileEntity> laboratoryResources = new HashMap<>();


    public MetaTileEntityIndustrialLaboratory(ResourceLocation metaTileEntityId) {
        super(metaTileEntityId, SerendustryRecipeMaps.LABORATORY_RECIPES);
        this.recipeMapWorkable = new IndustrialLaboratoryRecipeLogic(this);
    }

    @Override
    public MetaTileEntity createMetaTileEntity(IGregTechTileEntity iGregTechTileEntity) {
        return new MetaTileEntityIndustrialLaboratory(metaTileEntityId);
    }

    @Nonnull
    @Override
    protected BlockPattern createStructurePattern() {
        return FactoryBlockPattern.start()
                .aisle("XXXXX", "XGGGX", "XGGGX", "XXXXX")
                .aisle("XXXXX", "G###G", "G###G", "XXXXX")
                .aisle("XXXXX", "G###G", "G###G", "XXXXX")
                .aisle("XXXXX", "G###G", "G###G", "XXXXX")
                .aisle("XXSXX", "XGGGX", "XGGGX", "XXXXX")
                .where('S', selfPredicate())
                .where('X', states(getCasingState()).setMinGlobalLimited(14).or(autoAbilities()))
                .where('G', states(getGlassState()))
                .where('#', innerPredicate())
                .build();
    }

    @Override
    public ICubeRenderer getBaseTexture(IMultiblockPart iMultiblockPart) {
        return Textures.INERT_PTFE_CASING; // todo
    }

    protected IBlockState getCasingState() {
        // todo
        return MetaBlocks.MACHINE_CASING.getState(BlockMachineCasing.MachineCasingType.IV);
    }

    protected IBlockState getGlassState() {
        // todo
        return MetaBlocks.TRANSPARENT_CASING.getState(BlockGlassCasing.CasingType.LAMINATED_GLASS);
    }

    @Override
    public void invalidateStructure() {
        super.invalidateStructure();
        this.laboratoryResources.onInvalidateStructure();
    }

    @Nonnull
    protected TraceabilityPredicate innerPredicate() {
        return new TraceabilityPredicate(bws -> {
            // todo might want to be more allowing on what can be inside
            if (bws.getBlockState().getBlock().isAir(bws.getBlockState(), bws.getWorld(), bws.getPos())) {
                return true;
            }
            TileEntity tileEntity = bws.getTileEntity();
            if (!(tileEntity instanceof IGregTechTileEntity)) {
                // remove an existing resource if this somehow got removed without deforming the multiblock
                laboratoryResources.removePos(bws.getPos());
                return false;
            }

            MetaTileEntity metaTileEntity = ((IGregTechTileEntity) tileEntity).getMetaTileEntity();
            if (!LaboratoryProperty.isMachineAllowed(metaTileEntity) || !(metaTileEntity instanceof TieredMetaTileEntity))
                return false;
            TieredMetaTileEntity tieredMte = (TieredMetaTileEntity) metaTileEntity;

            // insert into resources map if it doesn't exist, or if the machine at position somehow got changed
            laboratoryResources.putAtPos(bws.getPos(), tieredMte);
            return true;
        });
    }

    @Override
    protected void formStructure(PatternMatchContext context) {
        super.formStructure(context);
        mode = Mode.NORMAL; // todo determine mode based on structure here
    }

    private enum Mode {
        NORMAL, PARALLEL_REACTOR, // todo more modes
    }

    private static class IndustrialLaboratoryRecipeLogic extends MultiblockRecipeLogic {

        public IndustrialLaboratoryRecipeLogic(RecipeMapMultiblockController tileEntity) {
            super(tileEntity);
        }

        @Nonnull
        @Override
        public MetaTileEntityIndustrialLaboratory getMetaTileEntity() {
            return (MetaTileEntityIndustrialLaboratory) super.getMetaTileEntity();
        }

        @Override
        public int getParallelLimit() {
            switch (getMetaTileEntity().mode) {
                case PARALLEL_REACTOR:
                    return 16; // todo details
                case NORMAL:
                default:
                    return 1;
            }
        }

        @Nullable
        @Override
        public RecipeMap<?> getRecipeMap() {
            switch (getMetaTileEntity().mode) {
                case PARALLEL_REACTOR:
                    return RecipeMaps.CHEMICAL_RECIPES;
                case NORMAL:
                default:
                    return SerendustryRecipeMaps.LABORATORY_RECIPES;
            }
        }

        @Override
        public boolean checkRecipe(@Nonnull Recipe recipe) {
            if (!super.checkRecipe(recipe)) {
                return false;
            }

            LaboratoryEntry entry = recipe.getProperty(LaboratoryProperty.getInstance(), null);
            if (entry != null) {
                return getMetaTileEntity().laboratoryResources.matchesEntry(entry);
            }
            return true;
        }
    }

    private static class LaboratoryResources {

        // placement data
        private final Map<BlockPos, TieredMetaTileEntity> internalMachines;

        private final Map<RecipeMap<?>, RecipeMapData> internalRecipeData;

        private LaboratoryResources() {
            this.internalMachines = new HashMap<>();
            this.internalRecipeData = new HashMap<>();
        }

        public void putAtPos(BlockPos pos, TieredMetaTileEntity machine) {
            if (!internalMachines.containsKey(pos) || internalMachines.get(pos) != machine) {
                RecipeMap<?> map = machine.getRecipeMap();
                int tier = machine.getTier();
                if (map == null || tier < 0) return;
                RecipeMapData mapData = internalRecipeData.computeIfAbsent(map, $ -> new RecipeMapData());
                mapData.insertAtTier(tier, 1);
                internalMachines.put(pos, machine);
            }
        }

        public void removePos(BlockPos pos) {
            TieredMetaTileEntity tieredMte = internalMachines.get(pos);
            if (tieredMte != null) {
                RecipeMap<?> map = tieredMte.getRecipeMap();
                int tier = tieredMte.getTier();
                if (map != null && tier >= 0) {
                    RecipeMapData mapData = internalRecipeData.get(map);
                    if (mapData != null) {
                        mapData.removeFromTier(tier, 1);
                    }
                }
                internalMachines.remove(pos);
            }
        }

        public void onInvalidateStructure() {
            internalMachines.clear();
            internalRecipeData.clear();
        }

        @SuppressWarnings("ConstantConditions")
        public boolean matchesEntry(LaboratoryEntry entry) {
            for (Table.Cell<RecipeMap<?>, Integer, Integer> cell : entry.getMachineTable().cellSet()) {
                RecipeMap<?> map = cell.getRowKey();
                int tier = cell.getColumnKey();
                int count = cell.getValue();
                RecipeMapData mapData = internalRecipeData.get(map);
                if (mapData.getNumMatchingTier(tier) < count) return false;
            }
            return true;
        }

        private static class RecipeMapData {

            // array of [tier] -> count
            private final int[] countArray;

            public RecipeMapData() {
                countArray = new int[GTValues.MAX];
            }

            public void insertAtTier(int tier, int count) {
                if (tier >= GTValues.MAX || tier < 0) return;
                for (int i = 0; i <= tier; i++) {
                    countArray[i] += count;
                }
            }

            public void removeFromTier(int tier, int count) {
                if (tier >= GTValues.MAX || tier < 0) return;
                for (int i = 0; i <= tier; i++) {
                    countArray[i] = Math.max(0, countArray[i] - count);
                }
            }

            public int getNumMatchingTier(int tier) {
                if (tier >= GTValues.MAX || tier < 0) return 0;
                return countArray[tier];
            }
        }
    }
}

File: src/main/kotlin/serendustry/machine/MetaTileEntityIndustrialLaboratory.kt
package serendustry.machine

import gregtech.api.GTValues
import gregtech.api.capability.impl.MultiblockRecipeLogic
import gregtech.api.metatileentity.TieredMetaTileEntity
import gregtech.api.metatileentity.interfaces.IGregTechTileEntity
import gregtech.api.metatileentity.multiblock.IMultiblockPart
import gregtech.api.metatileentity.multiblock.RecipeMapMultiblockController
import gregtech.api.pattern.BlockPattern
import gregtech.api.pattern.FactoryBlockPattern
import gregtech.api.pattern.PatternMatchContext
import gregtech.api.pattern.TraceabilityPredicate
import gregtech.api.recipes.Recipe
import gregtech.api.recipes.RecipeMap
import gregtech.api.recipes.RecipeMaps
import gregtech.client.renderer.ICubeRenderer
import gregtech.client.renderer.texture.Textures
import gregtech.common.blocks.BlockGlassCasing
import gregtech.common.blocks.BlockMachineCasing
import gregtech.common.blocks.MetaBlocks
import net.minecraft.block.state.IBlockState
import net.minecraft.util.ResourceLocation
import net.minecraft.util.math.BlockPos

class MetaTileEntityIndustrialLaboratory(id: ResourceLocation): RecipeMapMultiblockController(id, LABORATORY_RECIPES) {

    private var mode = Mode.NORMAL

    private val laboratoryResources = LaboratoryResources()

    init {
        this.recipeMapWorkable = IndustrialLaboratoryRecipeLogic(this)
    }

    override fun createMetaTileEntity(te: IGregTechTileEntity) = MetaTileEntityIndustrialLaboratory(metaTileEntityId)

    override fun createStructurePattern(): BlockPattern =
        FactoryBlockPattern.start()
            .aisle("XXXXX", "XGGGX", "XGGGX", "XXXXX")
            .aisle("XXXXX", "G###G", "G###G", "XXXXX")
            .aisle("XXXXX", "G###G", "G###G", "XXXXX")
            .aisle("XXXXX", "G###G", "G###G", "XXXXX")
            .aisle("XXSXX", "XGGGX", "XGGGX", "XXXXX")
            .where('S', selfPredicate())
            .where('X', states(getCasingState()).setMinGlobalLimited(14).or(autoAbilities()))
            .where('G', states(getGlassState()))
            .where('#', innerPredicate())
            .build()

    override fun getBaseTexture(part: IMultiblockPart?): ICubeRenderer = Textures.INERT_PTFE_CASING // todo

    // todo
    private fun getCasingState(): IBlockState = MetaBlocks.MACHINE_CASING.getState(BlockMachineCasing.MachineCasingType.IV)

    // todo
    private fun getGlassState(): IBlockState = MetaBlocks.TRANSPARENT_CASING.getState(BlockGlassCasing.CasingType.LAMINATED_GLASS)

    private fun innerPredicate(): TraceabilityPredicate = TraceabilityPredicate { bws ->
        if (bws.blockState.block.isAir(bws.blockState, bws.world, bws.pos)) {
            return@TraceabilityPredicate true
        }
        val tileEntity = bws.tileEntity
        if (tileEntity !is IGregTechTileEntity) {
            laboratoryResources.removePos(bws.pos)
            return@TraceabilityPredicate false
        }

        val mte = (tileEntity as IGregTechTileEntity).metaTileEntity
        if (!LaboratoryProperty.isMachineAllowed(mte) || mte !is TieredMetaTileEntity) {
            return@TraceabilityPredicate false
        }
        laboratoryResources.putAtPos(bws.pos, mte)
        return@TraceabilityPredicate true
    }

    override fun invalidateStructure() {
        super.invalidateStructure()
        laboratoryResources.onInvalidateStructure()
    }

    override fun formStructure(context: PatternMatchContext?) {
        super.formStructure(context)
        mode = Mode.NORMAL
    }

    private enum class Mode {
        NORMAL, PARALLEL_REACTOR, // todo more modes
    }

    private class IndustrialLaboratoryRecipeLogic(te: RecipeMapMultiblockController): MultiblockRecipeLogic(te) {

        override fun getMetaTileEntity() = super.getMetaTileEntity() as MetaTileEntityIndustrialLaboratory

        override fun getParallelLimit() = if (getMetaTileEntity().mode == Mode.PARALLEL_REACTOR) 16 else 1

        override fun getRecipeMap(): RecipeMap<*> = if (getMetaTileEntity().mode == Mode.PARALLEL_REACTOR)
            RecipeMaps.CHEMICAL_RECIPES else LABORATORY_RECIPES

        override fun checkRecipe(recipe: Recipe): Boolean {
            if (!super.checkRecipe(recipe)) return false
            val entry = recipe.getProperty(LaboratoryProperty.getInstance(), null) ?: return true
            return getMetaTileEntity().laboratoryResources.matchesEntry(entry)
        }
    }

    private class LaboratoryResources {

        private val internalMachines = mutableMapOf<BlockPos, TieredMetaTileEntity>()
        private val internalRecipeData = mutableMapOf<RecipeMap<*>, RecipeMapData>()

        fun putAtPos(pos: BlockPos, machine: TieredMetaTileEntity) {
            if (!internalMachines.containsKey(pos) || internalMachines[pos] != machine) {
                val map = machine.recipeMap ?: return
                val tier = machine.tier.takeUnless { it < 0 } ?: return
                internalRecipeData.computeIfAbsent(map) { RecipeMapData() }.also { it.insertAtTier(tier, 1) }
                internalMachines[pos] = machine
            }
        }

        fun removePos(pos: BlockPos) {
            val tieredMte = internalMachines[pos] ?: return
            val map = tieredMte.recipeMap ?: return
            val tier = tieredMte.tier
            internalRecipeData[map]?.takeIf { tier >= 0 }?.removeFromTier(tier, 1)
            internalMachines.remove(pos)
        }

        fun onInvalidateStructure() {
            internalMachines.clear()
            internalRecipeData.clear()
        }

        fun matchesEntry(entry: LaboratoryProperty.LaboratoryEntry): Boolean {
            for (cell in entry.getMachineTable().cellSet()) {
                val map = cell.rowKey ?: continue
                val tier = cell.columnKey ?: continue
                val count = cell.value ?: continue
                val mapData = internalRecipeData[map] ?: continue
                if (mapData.getNumMatchingTier(tier) < count) return false
            }
            return true
        }
    }

    private class RecipeMapData {

        private val countArray = IntArray(GTValues.MAX)

        fun insertAtTier(tier: Int, count: Int) {
            if (tier !in 0 until GTValues.MAX) return
            for (i in 0..tier) {
                countArray[i] += count
            }
        }

        fun removeFromTier(tier: Int, count: Int) {
            if(tier !in 0 until GTValues.MAX) return
            for (i in 0..tier) {
                countArray[i] = (countArray[i] - count).coerceAtLeast(0)
            }
        }

        fun getNumMatchingTier(tier: Int) = countArray.getOrNull(tier) ?: 0
    }
}
